### DISCUSSION TITLE
How to config field alias to take effect only in serialization and deserialization

### DISCUSSION THREAD

**Discussion by icefery** [pydantic/pydantic#9342] [Question] [answered]
How to config field alias to take effect only in serialization and deserialization?

For example:

```python
from pydantic import BaseModel, Field


class CustomObject(BaseModel):
    field_first: str = Field(alias="fieldFirst")
    field_second: str = Field(alias="fieldSecond")


if __name__ == "__main__":
    # deserialization
    json_data1 = """
{
    "fieldFirst": "a",
    "fieldSecond": "b"
}
    """
    obj1 = CustomObject.model_validate_json(json_data=json_data1)
    print("obj1", obj1, obj1.field_first, obj1.field_second)

    # serialization
    obj2 = CustomObject(field_first="a", field_second="b")
    json_data2 = obj2.model_dump_json(by_alias=True, indent=4)
    print("json_data_2", json_data2)
```

Since `snake_case` are commonly used in Python, and `camelCase` is commonly used elsewhere, it is expected that the alias will only take effect during serialization and deserialization.

Is there any way?

**Comment by asml-mdroogle** [comment]
https://docs.pydantic.dev/latest/api/config/#pydantic.config.ConfigDict.validate_by_name

**Comment by MukundaKatta** [comment]
Pydantic v2 splits this for you. Use `validation_alias` (incoming names) and `serialization_alias` (outgoing names) separately instead of the generic `alias`. Set only one to restrict to that direction. `AliasChoices` in `validation_alias` handles multiple incoming variants. That's the whole trick.

**Comment by x-tahosin** [comment]
In Pydantic v2 there are three separate knobs:

| Field arg | When it's used |
|---|---|
| `alias` | Both directions (input + output) |
| `validation_alias` | Input only (deserialization) |
| `serialization_alias` | Output only (serialization) |

So if you only want the alias on **one** side:

```python
from pydantic import BaseModel, Field

class CustomObject(BaseModel):
    field_first: str  = Field(validation_alias="fieldFirst")    # input only
    field_second: str = Field(serialization_alias="fieldSecond") # output only
```

But for your actual use case — *snake_case in Python, camelCase in JSON for both directions* — you don't want per-field aliases at all. The idiomatic way is the model-wide alias generator with `populate_by_name=True`:

```python
from pydantic import BaseModel, ConfigDict
from pydantic.alias_generators import to_camel

class CustomObject(BaseModel):
    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,    # also accept snake_case input
    )

    field_first: str
    field_second: str


# deserialize from camelCase
obj = CustomObject.model_validate_json('{"fieldFirst": "a", "fieldSecond": "b"}')

# serialize to camelCase  -> {"fieldFirst": "a", "fieldSecond": "b"}
print(obj.model_dump_json(by_alias=True))

# you can still construct with snake_case in Python
obj2 = CustomObject(field_first="a", field_second="b")
```

`populate_by_name=True` is what lets you keep using `field_first=...` in code while still accepting/emitting `fieldFirst` over the wire.

Refs:
- Aliases: https://docs.pydantic.dev/latest/concepts/alias/
- `alias_generators`: https://docs.pydantic.dev/latest/api/config/#pydantic.config.ConfigDict.alias_generator

**Comment by harshagarwalnyu** [comment, marked as answer]
In Pydantic v2, the `alias` parameter is a "catch-all" that applies to both validation and serialization. To get the specific behavior you're looking for—where the alias is used for external data interfaces but you remain in control of your Python-side naming—you should use `validation_alias` and `serialization_alias`.

### The Direct Solution

You can specify exactly when the alias should be used for each field:

```python
from pydantic import BaseModel, Field

class CustomObject(BaseModel):
    field_first: str = Field(validation_alias="fieldFirst", serialization_alias="fieldFirst")
    field_second: str = Field(validation_alias="fieldSecond", serialization_alias="fieldSecond")
```

### The "Best Practice" Solution (Global Config)

If you are dealing with a standard camelCase API (which is the most common use case for aliases), it is much more maintainable to use a global `alias_generator` in the `model_config`. This avoids duplicating the alias strings for every single field.

```python
from pydantic import BaseModel, ConfigDict
from pydantic.alias_generators import to_camel

class CustomObject(BaseModel):
    model_config = ConfigDict(
        alias_generator=to_camel,
        # This allows you to use snake_case in your Python constructor
        populate_by_name=True, 
    )
    
    field_first: str
    field_second: str

# Deserialization (Validation) - Accepts camelCase
obj = CustomObject.model_validate({"fieldFirst": "a", "fieldSecond": "b"})

# Internal Python usage - Use snake_case
print(obj.field_first)

# Serialization - Outputs camelCase
print(obj.model_dump(by_alias=True))
```

### Key Concepts:
1. **`validation_alias`**: Used only when reading data into the model (deserializing).
2. **`serialization_alias`**: Used only when outputting data from the model (serializing).
3. **`populate_by_name=True`**: This is crucial. It tells Pydantic: "Even though I have an alias for validation, still let me use the actual field name (`field_first`) when I'm creating the object in code." This is the key to having aliases that "only take effect" for the external boundary while staying idiomatic inside your Python logic.

**Comment by icefery** [comment]
Ok, thanks, `populate_by_name=True` option is the best way, cool!
